### ISSUE TITLE
Some scripts only download binary on Official website and it only can use on X86 rchitecture ,Can we support the ARM architecture?

### ISSUE THREAD

**Issue by luxianjun** [spack/spack#12212]
Hi.
I found that some packages are currently downloaded binary on Official website, but the binary is basically X86 architecture,I want to use on ARMarchitecture . So can we rewrite this kind of scripts to compile it from source code? For example, hadoop
Thanks!

**Comment by adamjstewart** [comment]
Another example of this is binary packages like `jdk`, where we always download the linux version, even on macOS. #2009 is the oldest discussion of this problem that I know of. An API I think would make sense would be something like:
```python
version('1.0', sha256=..., when='platform=linux')
version('1.0', sha256=..., when='platform=darwin')
```
or
```python
version('1.0', sha256=..., when='target=x86_64')
version('1.0', sha256=..., when='target=arm')
```
@tgamblin and I talked about this at one point, but it's non-trivial to change Spack to have multiple occurrences of the same version for different platforms.

**Comment by alalazo** [comment]
Can't we solve it already by overriding `url_for_version`? I need to double check the code but off the top of my head I'd say that function is called after concretization, so one could return different urls depending on the current target.

**Comment by adamjstewart** [comment]
`url_for_version` can swap URLs but not checksums.

**Comment by adamjstewart** [comment]
`hpcviewer` is another package that implemented something similar to what I did in #2009. This may be _good enough_ as long as the only difference between platforms is, for example, linux vs. darwin. Does someone want to do this for `jdk`/`openjdk`?

**Comment by mwkrentel** [comment]
Yes, for hpcviewer we use a binary distribution with different tar
files for x86_64, ppc64 and ppc64le, all with the same version, of
course.  We use something like:

```
if platform.machine() == 'x86_64':
    version(ver, url='...', sha256='...')
elif platform.machine() == 'ppc64':
    version(ver, url='...', sha256='...')
...
```

Actually, we use a table of version, machine type and SHA sum and a
`for` loop, but whatever.

This works for a property that you can test 'statically' from python,
outside of spack.  For example, Linux vs Macos, or x86 vs ppc, etc.
It doesn't work for testing something from a spec.

**Comment by alalazo** [comment]
Duplicate of #11206

**Comment by alalazo** [comment]
Closing as I feel it would be solved by what is proposed in #11206, feel free to reopen if I missed something.
